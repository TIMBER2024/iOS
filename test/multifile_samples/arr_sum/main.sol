pragma solidity 0.8.4;

contract Foo {
    // Simple map
    uint[] a = [10];
    int8[] b;

    /// #if_succeeds unchecked_sum(a) > 10 && unchecked_sum(a) < 20;
    function pushA(uint k) public {
        a.push(k);
    }

    /// #if_succeeds unchecked_sum(a) > 10 && unchecked_sum(a) < 20;
    function setA(uint k, uint v) public {
        a[k] = v;
    }

    /// #if_succeeds unchecked_sum(b) > -10 && unchecked_sum(b) < 10;
    function pushB(int8 k) public {
        b.push(k);
    }

    /// #if_succeeds unchecked_sum(b) > -10 && unchecked_sum(b) < 10;
    function setB(uint k, int8 v) public {
        b[k] = v;
    }
    

    /// #if_succeeds unchecked_sum(c) > -10 && unchecked_sum(c) < 10;
    function memArr(int16[] memory c) public {
    }

    /// #if_succeeds unchecked_sum(c) > -10 && unchecked_sum(c) < 10;
    function calldataArr(int16[] calldata c) external {
    }

    /// #if_succeeds unchecked_sum(c) < 10;
    function overflowCheck(uint[] calldata c) external {
    }
}
