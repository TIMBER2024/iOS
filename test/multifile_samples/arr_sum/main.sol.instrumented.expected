pragma solidity 0.8.4;

import "./__scribble_ReentrancyUtils.sol";

contract Foo {
    event AssertionFailed(string message);

    uint[] internal a = [10];
    int8[] internal b;

    function pushA(uint k) public {
        _original_Foo_pushA(k);
        unchecked {
            if (!((arr_sum_funs.sum_arr_uint256_arr_storage(a) > 10) && (arr_sum_funs.sum_arr_uint256_arr_storage(a) < 20))) {
                emit AssertionFailed("0: ");
                assert(false);
            }
        }
    }

    function _original_Foo_pushA(uint k) private {
        a.push(k);
    }

    function setA(uint k, uint v) public {
        _original_Foo_setA(k, v);
        unchecked {
            if (!((arr_sum_funs.sum_arr_uint256_arr_storage(a) > 10) && (arr_sum_funs.sum_arr_uint256_arr_storage(a) < 20))) {
                emit AssertionFailed("1: ");
                assert(false);
            }
        }
    }

    function _original_Foo_setA(uint k, uint v) private {
        a[k] = v;
    }

    function pushB(int8 k) public {
        _original_Foo_pushB(k);
        unchecked {
            if (!((arr_sum_funs.sum_arr_int8_arr_storage(b) > (-10)) && (arr_sum_funs.sum_arr_int8_arr_storage(b) < 10))) {
                emit AssertionFailed("2: ");
                assert(false);
            }
        }
    }

    function _original_Foo_pushB(int8 k) private {
        b.push(k);
    }

    function setB(uint k, int8 v) public {
        _original_Foo_setB(k, v);
        unchecked {
            if (!((arr_sum_funs.sum_arr_int8_arr_storage(b) > (-10)) && (arr_sum_funs.sum_arr_int8_arr_storage(b) < 10))) {
                emit AssertionFailed("3: ");
                assert(false);
            }
        }
    }

    function _original_Foo_setB(uint k, int8 v) private {
        b[k] = v;
    }

    function memArr(int16[] memory c) public {
        _original_Foo_memArr(c);
        unchecked {
            if (!((arr_sum_funs.sum_arr_int16_arr_memory(c) > (-10)) && (arr_sum_funs.sum_arr_int16_arr_memory(c) < 10))) {
                emit AssertionFailed("4: ");
                assert(false);
            }
        }
    }

    function _original_Foo_memArr(int16[] memory c) private {}

    function calldataArr(int16[] calldata c) external {
        _original_Foo_calldataArr(c);
        unchecked {
            if (!((arr_sum_funs.sum_arr_int16_arr_calldata(c) > (-10)) && (arr_sum_funs.sum_arr_int16_arr_calldata(c) < 10))) {
                emit AssertionFailed("5: ");
                assert(false);
            }
        }
    }

    function _original_Foo_calldataArr(int16[] calldata c) private {}

    function overflowCheck(uint[] calldata c) external {
        _original_Foo_overflowCheck(c);
        unchecked {
            if (!(arr_sum_funs.sum_arr_uint256_arr_calldata(c) < 10)) {
                emit AssertionFailed("6: ");
                assert(false);
            }
        }
    }

    function _original_Foo_overflowCheck(uint[] calldata c) private {}
}